lightning/src/chain/keysinterface.rs

```diff
@@ -187,6 +187,15 @@ impl_writeable_tlv_based_enum!(SpendableOutputDescriptor,
 	(2, StaticPaymentOutput),
 );
 
+/// An error while signing
+#[derive(Debug)]
+pub enum SignError {
+	/// The signer is temporarily unavailable
+	Temporary,
+	/// A signer internal error
+	Internal
+}
```

Comment:
> Just to confirm if we agree all on the direction, in case of `SignError::Temporary` yielded by `finalize_claim_tx`, we return a `PendingEvent::SignTransaction` to `ChannelMonitor` queued on a new internal buffer ?

Initially, I thought we would just call `ChannelMonitor.update_claims_view` with empty inputs and the current logic would have it retry the requests in `self.pending_claim_requests`.  It seemed like the least invasive approach.

However, it seems that @TheBlueMatt prefers if we just retry the signing/broadcasting part, and not the rest of the logic.  Maybe he can clarify how his envisions the triggering of the retry.

As to implementation details - there are already three buffers in `ChannelMonitorImpl`:

```rust
	pending_monitor_events: Vec<MonitorEvent>,
	pending_events: Vec<Event>,
	onchain_events_awaiting_threshold_conf: Vec<OnchainEventEntry>,
```

I'm wondering if any of these could be used for this purpose.

> 
> How we would dry-up the queue ? Inside `block_confirmed` or with a new `monitor_timer_tick_occured()`, I lean to favor the latter as blocks elapsing is a bit precious in Lightning.

The signer may connect at any time, which may not overlap with the block arrival time.  So it's probably best that the `BackgroundProcessor` tick gets triggered by the signer connecting so we can catch up on all housekeeping tasks.  `ChannelMonitor` already implements `EventsProvider::process_pending_events`, so perhaps it makes sense to add the code there.  Or it may be cleaner to add another call as you say.

> 
> If we get an `SignError::Internal` maybe we should flow it back up to `ChannelManager` with a `ChannelMonitor::TemporaryFailure` and stop accepting HTLCs we won't be able to claim anyways ?

The `Internal` error is just a renaming of the existing `Result<_, ()>` error returns from `BaseSign`.  It seems like we currently almost always panic, or we should, because it's caused by a key derivation failing, which has an infinitesimal chance of happening and probably indicates a hardware / RNG problem.  In any case, it's out of scope of the current effort to change the handling of this case.

Reply:
> I'd imagine we'd leave that up to the user or maybe even do it on every block (though that will basically break ~every test, so maybe we don't do that by default in tests?).

IIUC the idea as just retrying the signing/broadcasting part I'm good with that, it's more likely the `update_claims_view` logic is going to be more complex in the future (e.g locking fee-bumping UTXOs) though to break it to avoid dependency bug like the one above.

> As to implementation details - there are already three buffers in ChannelMonitorImpl:

I think the last one `onchain_events_awaiting_threshold_conf` would fit, all the events are destinied to be consumed by `ChannelMonitorImpl`. Maybe can be recall something like "monitor_events_awaiting_time_point` where time point is either a blockchain height or the signer waking up.

> The signer may connect at any time, which may not overlap with the block arrival time. So it's probably best that the BackgroundProcessor tick gets triggered by the signer connecting so we can catch up on all housekeeping tasks. ChannelMonitor already implements EventsProvider::process_pending_events, so perhaps it makes sense to add the code there. Or it may be cleaner to add another call as you say.

Yeah `ChannelMonitorImpl::get_and_clear_pending_events` sounds a good temporary location for now. 

> The Internal error is just a renaming of the existing Result<_, ()> error returns from BaseSign. It seems like we currently almost always panic, or we should, because it's caused by a key derivation failing, which has an infinitesimal chance of happening and probably indicates a hardware / RNG problem. In any case, it's out of scope of the current effort to change the handling of this case.

Sure, we can deferred how we react to catastrophic hardware failures to the future.